course_inventory/published_date.py

```diff
@@ -75,40 +106,113 @@ def published_date_resp_parsing(self, response):
+                    self.published_date_retry_bucket.pop(course_id)
+                    logger.info(f"Removed {course_id} removed from retry list {self.published_date_retry_bucket}")
                 break
         if not published_date_found:
             self.get_next_page_url(response)
 
-        seconds = time.time() - start_time
-        str_time = time.strftime("%H:%M:%S", time.gmtime(seconds))
-        logger.debug(f"Parsing the published date took {str_time} ")
         return
 
-    def get_published_course_date(self, course_ids, next_page_links=None):
-        logger.info("Starting of get_published_course_date call")
-        with FuturesSession(max_workers=self.num_workers) as session:
+    def retry_logic_with_error(self, course_id, url) -> None:
+        if course_id in self.published_date_retry_bucket:
+            retry_count = self.published_date_retry_bucket[course_id]['count']
+            if retry_count > self.retry_attempts - 1:
+                # Don't want to retry more
+                logger.info(f"Reached max retry attempts for {course_id} with count {retry_count}, Removing... ")
+                self.published_date_retry_bucket.pop(course_id)
+            else:
+                # We will give one more chance to retry
+                logger.info(f"Retry more for {course_id}  ")
+                self.published_date_retry_bucket[course_id]['count'] += 1
+        else:
+            logger.info(f"Putting course {course_id} to retry list")
+            self.published_date_retry_bucket[course_id] = {
+                'url': url,
+                'count': 1,
+            }
+        logger.info(f"Retry list size {len(self.published_date_retry_bucket)} for published_at date")
+
+    def filter_courses_to_fetch_published_date(self) -> pd.DataFrame:
+        pd.set_option('display.max_column', None)
+        logger.info(f"Size of courses data from API routine: {self.course_data_from_api.shape}")
+        state_available_courses_from_api = self.course_data_from_api[
+            (self.course_data_from_api['workflow_state'] == 'available')].shape
+        logger.info(f"""Size of Published courses from API: {state_available_courses_from_api}""")
+        logger.info(f"Size of published courses from DB with: {self.course_data_from_db.shape}")
+        published_date_in_db = self.course_data_from_db[(self.course_data_from_db['published_at'].notnull())].shape
+        logger.info(f"Size of published courses from DB with published date: {published_date_in_db}")
+        course_with_pub_date_added_from_df = pd.merge(self.course_data_from_api, self.course_data_from_db,
+                                                      on='canvas_id', how='left')
+        logger.info(f"Size of course data after merging with DB data: {course_with_pub_date_added_from_df.shape}")
+        return course_with_pub_date_added_from_df
+
+    def get_published_course_date(self, course_ids, retry_list=None) -> None:
```

Comment:
Nit-pick: The arguments here could use type hints for clarity. `retry_list` is confusing since I think the outermost data type is `Dict`?

Reply:
I have added the type hint

Reply:
Done